File: backend/app/api/routers/challenge.py
from app.service.user import UserService
from app.helper.auth import get_current_user
from app.schema.challenge_schema import ChallengeCreate, ChallengeOut
from app.service.challenge import ChallengeService
from fastapi import APIRouter, Depends, HTTPException
from app.db.connect import get_db

router = APIRouter(prefix="/challenges", tags=["Challenges"])

@router.get("/scene/{scene_id}", response_model=list[ChallengeOut])
async def list_challenges_for_scene(scene_id: str, db=Depends(get_db)):
    return await ChallengeService.list_challenges_for_scene(scene_id, db)

@router.get("/{challenge_id}", response_model=ChallengeOut)
async def get_challenge(challenge_id: str, db=Depends(get_db)):
    return await ChallengeService.get_challenge(challenge_id, db)

@router.post("/", response_model=ChallengeOut)
async def create_challenge(body: ChallengeCreate, db=Depends(get_db)):
    return await ChallengeService.create_challenge(body, db)

@router.put("/{challenge_id}", response_model=ChallengeOut)
async def update_challenge(challenge_id: str, body: ChallengeCreate, db=Depends(get_db)):
    return await ChallengeService.update_challenge(challenge_id, body, db)

@router.delete("/{challenge_id}")
async def delete_challenge(challenge_id: str, db=Depends(get_db)):
    return await ChallengeService.delete_challenge(challenge_id, db)

@router.post("/{challenge_id}/attempt")
async def submit_challenge_attempt(challenge_id: str, story_id: str, answer_given: str, current_user=Depends(get_current_user), db=Depends(get_db)):
    return await ChallengeService.submit_attempt(current_user["id"], challenge_id, story_id, answer_given, db)    

@router.get("/{challenge_id}/attempts")
async def get_user_attempts(challenge_id: str, story_id: str, current_user=Depends(get_current_user), db=Depends(get_db)):
    return await ChallengeService.get_user_attempts(current_user["id"], challenge_id, db)

File: frontend/api/challenge.ts
import api from "./api";

export type QuestionType = "quiz" | "puzzle" | "riddle" | "matching";

export interface Challenge {
  id: string;
  scene_id: string;
  type: QuestionType;
  options: Record<string, unknown> | null;
  correct_answer: string;
  max_attempt: number;
  created_at: string;
}

export interface CreateChallengePayload {
  scene_id: string;
  type: QuestionType;
  options: Record<string, unknown> | null;
  correct_answer: string;
  max_attempt: number;
}

export async function listChallengesForScene(sceneId: string): Promise<Challenge[]> {
  const { data } = await api.get<Challenge[]>(`/challenges/scene/${sceneId}`);
  return data;
}

export async function getChallenge(challengeId: string): Promise<Challenge> {
  const { data } = await api.get<Challenge>(`/challenges/${challengeId}`);
  return data;
}

export async function createChallenge(payload: CreateChallengePayload): Promise<Challenge> {
  const { data } = await api.post<Challenge>("/challenges/", payload);
  return data;
}

export async function updateChallenge(challengeId: string, payload: CreateChallengePayload): Promise<Challenge> {
  const { data } = await api.put<Challenge>(`/challenges/${challengeId}`, payload);
  return data;
}

export async function deleteChallenge(challengeId: string): Promise<void> {
  await api.delete(`/challenges/${challengeId}`);
}

export async function submitChallengeAttempt(challengeId: string, storyId: string, answerGiven: string): Promise<void> {
  await api.post(`/challenges/${challengeId}/attempt`, null, {
      params: { story_id: storyId, answer_given: answerGiven,},
    });
}

export async function getUserAttempts(challengeId: string, storyId: string): Promise<unknown> {
  const { data } = await api.get(`/challenges/${challengeId}/attempts`, {
    params: { story_id: storyId }
  });
  return data;
}